pkg/registry/clusteroperator/clusterversion/strategy.go

```diff
@@ -0,0 +1,124 @@
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/apiserver/pkg/storage/names"
+
+	"github.com/golang/glog"
+	"github.com/openshift/cluster-operator/pkg/apis/clusteroperator"
+	"github.com/openshift/cluster-operator/pkg/apis/clusteroperator/validation"
+)
+
+// NewScopeStrategy returns a new NamespaceScopedStrategy for clusters
+func NewScopeStrategy() rest.NamespaceScopedStrategy {
+	return clusterRESTStrategies
+}
+
+// implements interfaces RESTCreateStrategy, RESTUpdateStrategy, RESTDeleteStrategy,
+// NamespaceScopedStrategy
+type clusterRESTStrategy struct {
+	runtime.ObjectTyper // inherit ObjectKinds method
+	names.NameGenerator // GenerateName method for CreateStrategy
+}
+
+// implements interface RESTUpdateStrategy
+type clusterVersionStatusRESTStrategy struct {
+	clusterRESTStrategy
+}
+
+var (
+	clusterRESTStrategies = clusterRESTStrategy{
+		// embeds to pull in existing code behavior from upstream
+
+		// this has an interesting NOTE on it. Not sure if it applies to us.
+		ObjectTyper: api.Scheme,
+		// use the generator from upstream k8s, or implement method
+		// `GenerateName(base string) string`
+		NameGenerator: names.SimpleNameGenerator,
+	}
+	_ rest.RESTCreateStrategy = clusterRESTStrategies
+	_ rest.RESTUpdateStrategy = clusterRESTStrategies
+	_ rest.RESTDeleteStrategy = clusterRESTStrategies
+
+	clusterVersionStatusUpdateStrategy = clusterVersionStatusRESTStrategy{
+		clusterRESTStrategies,
+	}
+	_ rest.RESTUpdateStrategy = clusterVersionStatusUpdateStrategy
+)
+
+// Canonicalize does not transform a cluster.
+func (clusterRESTStrategy) Canonicalize(obj runtime.Object) {
+	_, ok := obj.(*clusteroperator.ClusterVersion)
+	if !ok {
+		glog.Fatal("received a non-cluster version object to create")
+	}
+}
+
+// NamespaceScoped returns true as cluster versions are scoped to a namespace.
+func (clusterRESTStrategy) NamespaceScoped() bool {
+	return true
+}
+
+func (clusterRESTStrategy) PrepareForCreate(ctx genericapirequest.Context, obj runtime.Object) {
+	_, ok := obj.(*clusteroperator.ClusterVersion)
+	if !ok {
+		glog.Fatal("received a non-cluster version object to create")
+	}
+}
+
+func (clusterRESTStrategy) Validate(ctx genericapirequest.Context, obj runtime.Object) field.ErrorList {
+	return validation.ValidateClusterVersion(obj.(*clusteroperator.ClusterVersion))
+}
+
+func (clusterRESTStrategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+func (clusterRESTStrategy) AllowUnconditionalUpdate() bool {
+	return false
+}
+
+func (clusterRESTStrategy) PrepareForUpdate(ctx genericapirequest.Context, new, old runtime.Object) {
+	_, ok := new.(*clusteroperator.ClusterVersion)
+	if !ok {
+		glog.Fatal("received a non-cluster version object to update to")
+	}
+	_, ok = old.(*clusteroperator.ClusterVersion)
+	if !ok {
+		glog.Fatal("received a non-cluster version object to update from")
+	}
+}
+
+func (clusterRESTStrategy) ValidateUpdate(ctx genericapirequest.Context, new, old runtime.Object) field.ErrorList {
```

Comment:
Need to add a `PrepareForUpdate` and `ValidateUpdate` for `clusterVersionStatusRESTStrategy`.